Setting permissions so that this container will work in OpenShift

secondsun opened the pull request:
I am basing this PR off of matzew's PR here : https://github.com/jboss-dockerfiles/wildfly/pull/42
I've tested it on the 3.6 alpha release of Origin by using the Dockerfile in a source build.

goldmann commented on server/Dockerfile:
```diff
@@ -13,6 +13,16 @@ USER jboss
+RUN mkdir -p /opt/jboss/keycloak/standalone/log
+RUN mkdir -p /opt/jboss/keycloak/standalone/data
+RUN chown -R jboss:0 /opt/jboss/keycloak/standalone
+RUN chmod -R g+rw /opt/jboss/keycloak/standalone
```
It looks good to me, but please combine all four RUN calls into one.

stianst replied:
Is it not sufficient to just set the correct permissions recursively on /opt/jboss/keycloak? It seems a bit strange that the log and data dirs have to be created at this point.

goldmann replied:
Yes, it is sufficient. Generally speaking $JBOSS_HOME probably can be used.

stianst replied:
I'd prefer that approach as it looks a lot less hacky ;)

Vad1mo replied:
No offense, but to me it looks like someone is building a docker images for the first time.

secondsun replied:
@Vad1mo Thanks, that is helpful.

secondsun replied:
@stianst It does look a little weird, but I took the approach of doing the least number changes necessary to make it work.  These are the directories OpenShift was barking about, though making it $JBOSS_HOME and seeing what happens is an option...

Vad1mo commented:
Updating permission so it runs on OpenShift is one thing, another is to have a openshift specific image. 
That contains config for a (HTTP/HTTPS) reverse proxy. Would another PR make sense?

secondsun commented:
@Vad1mo I can include the reverse proxy bits in this PR as well as I've had to implement them before in other situations.

secondsun commented:
Fixed formatting, used $JBOSS_HOME (and added it where it wasn't used) and added proxy-address-formatting so you can now use Edge encryption in OpenShift out of the box

Vad1mo commented:
@secondsun thanks, also for the proxy. 

One Question the  Proxy address forwarding changes, won't they affect negatively the normal user that are not on openshift or without a reverse proxy in the front?  

What do you say guys @goldmann @stianst?

secondsun commented:
@Vad1mo So the problem with just leaving the "proxy-address-forwarding="true"" setting is that the proxy headers could be added to make the application misbehave.  

Giving it a little bit of thought we probably shouldn't have this changed by default because the naive assumption is that Docker images use the system defaults and variations are exposed via environment variables.

So I guess it might make more sense to create a REVERSE_PROXY environment variable set to false and have the standalone.xml file reference that variable when it starts.  Meanwhile we document it so when it comes time to use it people can just say "-e REVERSE_PROXY=true" when they fire up their docker container if they need it.

wdyt?

Vad1mo commented:
yes this is the right direction. env var switches are common practice that cover the most frequent cases before you have to mount config file or create a custom image. Can jboss/keyclock deal with env vars in config files or are 3rd party tools needed?

stianst commented:
We can't have proxy-address-forwarding="true" enabled by default as that is a vulnerability if there is no actual load balancer/reverse proxy.

There'a also clustering to consider. On OpenShift it's very easy to spin up multiple instances and that's kinda the whole point with Kubernetes. However, this image doesn't deal with clustering and it certainly won't deal with clustering the way it has to be done on OpenShift.

In general I'm not really for adding a whole bunch of band aids to this image to get it to work on OpenShift. I'd rather see a properly developed OpenShift specific image including the templates and such to boot it up on OpenShift.

secondsun commented:
So the reverse-proxy isn't a hack just for OpenShift.  My SOP pre open shift for personal projects was to fire up docker containers behind an apache proxy that handled SSL.

As far as the changing directory ownership, given your argument w.r.t. clustering are you saying you would rather the not clusterable image not work at all in OpenShift (which pretty much nixes this PR) or that PRs for adding OpenShift features should go to an OpenShift image?

stianst commented:
The reverse proxy option is a nice add on and I'm happy to include that. I'd prefer it as a separate clean PR though.

In general I'm worried that to have something working fully on OpenShift there are more and more "hacks" being brought into the image. In the end I'm convinced that for a proper solution a separate image is going to be needed in the long run.

Medium/long term the hope is that we can provide a proper solution for OpenShift based off the RH-SSO image provided by cloud enablement.

In the short term the better option might be to add server-openshift that extends this image to add OpenShift specific things like file permissions, clustering, logging, etc..

secondsun commented:
@stianst  So how about this I 1) rework the reverse proxy into a separate PR and have it switchable as a Environment variable (defaulted to false) and 2) rework this PR to be a Dockerfile at server-openshift/Dockerfile which extends from the "base" PR.  

The biggest downside to #2 is that now in order to use keycloak-postgres I will need to make a keycloak-openshift-postgres image which will mostly be a copy of the original that inherits from a keycloak-openshift container.

stianst commented:
Different images for different databases is a bit of a pain. Would be better if the main server image just had support for different dbs. As it needs env variables for the db in either case it could just have an extra env variable for the db type. So you'd do:

    docker run jboss/keycloak -e DB=mysql ....

It's a fairly big change, but probably worth it.

stianst commented:
I love the DB idea and that's how it should have been all along, so I reckon I'm going to do that ;)

secondsun commented:
@stianst Ok, The permission changes are in server-openshift now and I opened a new PR for the proxy changes.

secondsun commented:
Rebased and cleaned up commit log

stianst commented:
Thanks

mguillem commented:
As of Keycloak 3.2.0.Final, the problem still persist.

This is easily reproducible: 
```
oc new-app jboss/keycloak:3.2.0.Final
```

the pods fail to start with following exception:
```
java.lang.IllegalArgumentException: Failed to instantiate class "org.jboss.logmanager.handlers.PeriodicRotatingFileHandler" for handler "FILE"
	at org.jboss.logmanager.config.AbstractPropertyConfiguration$ConstructAction.validate(AbstractPropertyConfiguration.java:116)
	at org.jboss.logmanager.config.LogContextConfigurationImpl.doPrepare(LogContextConfigurationImpl.java:335)
	at org.jboss.logmanager.config.LogContextConfigurationImpl.prepare(LogContextConfigurationImpl.java:288)
	at org.jboss.logmanager.config.LogContextConfigurationImpl.commit(LogContextConfigurationImpl.java:297)
	at org.jboss.logmanager.PropertyConfigurator.configure(PropertyConfigurator.java:546)
	at org.jboss.logmanager.PropertyConfigurator.configure(PropertyConfigurator.java:97)
	at org.jboss.logmanager.LogManager.readConfiguration(LogManager.java:514)
	at org.jboss.logmanager.LogManager.readConfiguration(LogManager.java:476)
	at java.util.logging.LogManager$3.run(LogManager.java:399)
	at java.util.logging.LogManager$3.run(LogManager.java:396)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.util.logging.LogManager.readPrimordialConfiguration(LogManager.java:396)
	at java.util.logging.LogManager.access$800(LogManager.java:145)
	at java.util.logging.LogManager$2.run(LogManager.java:345)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.util.logging.LogManager.ensureLogManagerInitialized(LogManager.java:338)
	at java.util.logging.LogManager.getLogManager(LogManager.java:378)
	at org.jboss.modules.Main.main(Main.java:482)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.jboss.logmanager.config.AbstractPropertyConfiguration$ConstructAction.validate(AbstractPropertyConfiguration.java:114)
	... 17 more
Caused by: java.io.FileNotFoundException: /opt/jboss/keycloak/standalone/log/server.log (No such file or directory)
	at java.io.FileOutputStream.open0(Native Method)
	at java.io.FileOutputStream.open(FileOutputStream.java:270)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
	at org.jboss.logmanager.handlers.FileHandler.setFile(FileHandler.java:151)
	at org.jboss.logmanager.handlers.PeriodicRotatingFileHandler.setFile(PeriodicRotatingFileHandler.java:102)
	at org.jboss.logmanager.handlers.FileHandler.setFileName(FileHandler.java:189)
	at org.jboss.logmanager.handlers.FileHandler.<init>(FileHandler.java:119)
	at org.jboss.logmanager.handlers.PeriodicRotatingFileHandler.<init>(PeriodicRotatingFileHandler.java:70)
	... 22 more
java.lang.IllegalStateException: WFLYSRV0124: Could not create server data directory: /opt/jboss/keycloak/standalone/data
	at org.jboss.as.server.ServerEnvironment.<init>(ServerEnvironment.java:473)
	at org.jboss.as.server.Main.determineEnvironment(Main.java:297)
	at org.jboss.as.server.Main.main(Main.java:94)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.modules.Module.run(Module.java:329)
at org.jboss.modules.Main.main(Main.java:507)
```

I think that `/opt/jboss/keycloak` has to be world writable whereas the applied patch makes it only group writable.

What I'm using:
```
FROM jboss/keycloak:3.2.0.Final

USER root
RUN chmod -R o+rw /opt/jboss/keycloak
```

Vad1mo commented:
@mguillem The image `jboss/keycloak:3.2.0.Final` does not has this entry
```
USER root

#Give correct permissions when used in an OpenShift environment.
RUN chown -R jboss:0 $JBOSS_HOME/standalone && \
    chmod -R g+rw $JBOSS_HOME/standalone

USER jboss
```

Just add this to your dockerfile and you are ready to go.

mguillem commented:
@Vad1mo I agree, this works as well.

Why aren't these settings used for the standard image jboss/keycloak:3.2.0.Final? Or why isn't a "jboss/keycloak-openshift:3.2.0.Final" pushed?

secondsun commented:
@mguillem See the discussion in this thread beginning here : https://github.com/jboss-dockerfiles/keycloak/pull/68#issuecomment-296925169 to answer your questions.

TL;DR; if there is no OpenShift image until there is OpenShift style clustering support.

mguillem commented:
@secondsun thanks for the link. Indeed slowly too long.

My point here: making the folders writeable allows to simply configure the rest within Openshift by using a dockerStrategy for the build config. We had our custom theme as well as the db driver this way and the standalone.xml is set using a volume from a config map. Only fixing the write rights can't be done in OpenShift (or at least with the normal security context).
